src/libopensc/card-idprime.c

```diff
@@ -0,0 +1,698 @@
+
+	LOG_FUNC_CALLED(card->ctx);
+	sc_log(card->ctx, "cmd=%ld ptr=%p", cmd, ptr);
+
+	if (priv == NULL) {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	}
+	switch (cmd) {
+		case SC_CARDCTL_GET_SERIALNR:
+			return idprime_get_serial(card, (sc_serial_number_t *) ptr);
+		case SC_CARDCTL_IDPRIME_INIT_GET_OBJECTS:
+			return idprime_get_init_and_get_count(&priv->pki_list, &priv->pki_current,
+				(int *)ptr);
+		case SC_CARDCTL_IDPRIME_GET_NEXT_OBJECT:
+			return idprime_fill_prkey_info(&priv->pki_list, &priv->pki_current,
+				(sc_pkcs15_prkey_info_t *)ptr);
+		case SC_CARDCTL_IDPRIME_FINAL_GET_OBJECTS:
+			return idprime_final_iterator(&priv->pki_list);
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+}
+
+#define HEADER_LEN 4
+
+static int idprime_select_file(sc_card_t *card, const sc_path_t *in_path, sc_file_t **file_out)
+{
+	int r, len;
+	idprime_private_data_t * priv = card->drv_data;
+	u8 data[HEADER_LEN];
+	size_t data_len = HEADER_LEN;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (priv) { /* don't record anything if we haven't been initialized yet */
+		/* forget any old cached values */
+		if (priv->cache_buf) {
+			free(priv->cache_buf);
+			priv->cache_buf = NULL;
+		}
+		priv->cache_buf_len = 0;
+		priv->cached = 0;
+	}
+	r = iso_ops->select_file(card, in_path, file_out);
+	if (r == SC_SUCCESS && priv && file_out != NULL) {
+		/* Try to read first bytes of the file to fix FCI in case of
+		 * compressed certififcate */
+		len = iso_ops->read_binary(card, 0, data, data_len, 0);
+		if (len == HEADER_LEN && data[0] == 0x01 && data[1] == 0x00) {
+			/* Cache the real file size for the caching read_binary() */
+			priv->file_size = (*file_out)->size;
+			/* Fix the information in the file structure to not confuse upper layers */
+			(*file_out)->size = (data[3]<<8) | data[2];
+		}
+	}
+	/* Return the exit code of the select command */
+	return r;
+}
+
+// used to read existing certificates
+static int idprime_read_binary(sc_card_t *card, unsigned int offset,
+	unsigned char *buf, size_t count, unsigned long flags)
+{
+	struct idprime_private_data *priv = card->drv_data;
+	int r;
+	int size;
+
+	sc_log(card->ctx, "called; %"SC_FORMAT_LEN_SIZE_T"u bytes at offset %d",
+	       count, offset);
+
+	if (!priv->cached && offset == 0) {
+		// this function is called to read and uncompress the certificate
+		u8 buffer[SC_MAX_EXT_APDU_BUFFER_SIZE];
+		if (sizeof(buffer) < count) {
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+		}
+		/* Read what was reported by FCI from select command */
+		r = iso_ops->read_binary(card, 0, buffer, priv->file_size, flags);
+		if (r < 0) {
+			LOG_FUNC_RETURN(card->ctx, r);
+		}
+		if (r < 4) {
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_DATA);
+		}
+		if (buffer[0] == 1 && buffer[1] == 0) {
+#ifdef ENABLE_ZLIB
+			size_t expectedsize = buffer[2] + buffer[3] * 0x100;
+			r = sc_decompress_alloc(&priv->cache_buf, &(priv->cache_buf_len),
```

Comment:
Could you have a look at sc_decompress_zlib_alloc? It looks like there's are multiple problems with
- calling realloc with size 0 (possible double free)
- missing length checks when calculating the buffer sizes
- possible misconception of realloc (afaik, "shrinking down" a buffer is not possible)

This could eventually hit you when passing cache_buf and cache_buf_len without any checks.

Reply:
I think the `cache_buf` member is freed when different applet is selected and therefore we should never hit any issues with realloc. It just gets the buffer with compressed certificate and decompress that into the `cache_buf`, which is allocated by `sc_decompress_zlib_alloc()`.

Reply:
My comment was not about your code, but about the implementation of `sc_decompress_alloc()` itself. Maybe this is better treated in a seperate issue....

I've opened https://github.com/OpenSC/OpenSC/issues/1905